x-pack/plugins/security_solution/server/lib/detection_engine/signals/mappings/build_risk_score_from_mapping.test.ts

```diff
@@ -4,23 +4,218 @@
+      });
+    });
+
+    test(`returns default score if the "number" is > 100`, () => {
+      testIt({
+        fieldValue: '101',
+        scoreDefault: 57,
+        scoreMapping: mappingToSingleField(),
+        expected: scoreOf(57),
+      });
+    });
+  });
 
-    expect(riskScore).toEqual({ riskScore: 57, riskScoreMeta: {} });
+  describe('base cases: when mapping to an array of numbers or strings', () => {
+    test(`returns that number if it's a single element and it's within the range [0;100]`, () => {
+      testIt({
+        fieldValue: [3.14],
+        scoreDefault: 57,
+        scoreMapping: mappingToSingleField(),
+        expected: overridenScoreOf(3.14),
+      });
+    });
+
+    test(`returns the max number of those that are within the range [0;100]`, () => {
+      testIt({
+        fieldValue: [42, -42, 17, 87, 87.5, '86.5', 110, 66],
+        scoreDefault: 57,
+        scoreMapping: mappingToSingleField(),
+        expected: overridenScoreOf(87.5),
+      });
+    });
+
+    test(`supports casting strings to numbers`, () => {
+      testIt({
+        fieldValue: [-1, 1, '3', '1.5', '3.14', 2],
+        scoreDefault: 57,
+        scoreMapping: mappingToSingleField(),
+        expected: overridenScoreOf(3.14),
+      });
+    });
   });
 
-  // TODO: Enhance...
+  describe('edge cases: when mapping to a single junk value', () => {
+    describe('ignores it and returns the default score', () => {
+      const cases = [
+        undefined,
+        null,
+        NaN,
+        Infinity,
+        -Infinity,
+        Number.MAX_VALUE,
+        -Number.MAX_VALUE,
+        -Number.MIN_VALUE,
+        'string',
+        [],
+        {},
+        new Date(),
+      ];
+
+      test.each(cases)('%p', (value) => {
+        testIt({
+          fieldValue: value,
+          scoreDefault: 57,
+          scoreMapping: mappingToSingleField(),
+          expected: scoreOf(57),
+        });
+      });
+    });
+  });
+
+  describe('edge cases: when mapping to an array of junk values', () => {
+    describe('ignores junk, extracts valid numbers and returns the max number within the range [0;100]', () => {
+      type Case = [unknown[], number];
+      const cases: Case[] = [
+        [[undefined, null, 1.5, 1, -Infinity], 1.5],
+        [['42', NaN, '44', '43', 42, {}], 44],
+        [[Infinity, '101', 100, 99, Number.MIN_VALUE], 100],
+        [[Number.MIN_VALUE, -0], Number.MIN_VALUE],
+      ];
+
+      test.each(cases)('%p', (value, expectedScore) => {
+        testIt({
+          fieldValue: value,
+          scoreDefault: 57,
+          scoreMapping: mappingToSingleField(),
+          expected: overridenScoreOf(expectedScore),
+        });
+      });
+    });
+  });
 });
+
+interface TestCase {
+  fieldValue: unknown;
+  scoreDefault: RiskScore;
+  scoreMapping: RiskScoreMappingOrUndefined;
+  expected: BuildRiskScoreFromMappingReturn;
+}
+
+function testIt({ fieldValue, scoreDefault, scoreMapping, expected }: TestCase) {
```

Comment:
Thanks, I briefly went through these docs:

- https://www.elastic.co/guide/en/kibana/master/pr-review.html
- https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md
- https://github.com/elastic/kibana/blob/master/TYPESCRIPT.md

Reply:
On functions, arrows vs fn expressions vs fn declarations. Normally, unless there's a different coding standard/guideline, to vertically structure the code I prefer using the ["newspaper metaphor"](https://henrikwarne.com/2015/01/03/book-review-clean-code/) coined by Uncle Bob in his Clean Code book.

> Newspaper Metaphor. The source file should be organized like a newspaper article, with the highest level summary at the top, and more and more details further down. Functions called from the top function come directly below it, and so on down to the lowest level and most detailed functions at the bottom. This is a good way to organize the source code, even though IDE:s make the location of functions less important, since it is so easy to navigate in and out of them.

It's a top-down approach, from general things to details, and it's possible with function declarations. With arrows and fn expressions it's possible to use the opposite bottow-up approach, where you first have details, helper functions etc, and at the bottom of the file there's let's say the main function.

Normally I use arrows in all other cases: inline calls, closures to an outer this, just a bunch of small independent functions in a file (e.g. selectors), etc.

Of course I wouldn't go against our guidelines, whether they are documented or not, and against the typical style of the codebase. For the sake of consistency.

Hopefully I managed to explain my motivation. Let me know what you think. Although the [styleguide](https://github.com/elastic/kibana/blob/master/STYLEGUIDE.md) uses fn declarations, it's a general one. If this style looks too foreign for the `security_solution` plugin's codebase, I'll change it to the one based on arrows.

Reply:
On the `testIt` helper. Yea, it definitely has pros and cons. It reduces duplication and makes the test suite shorter, but abstracts away some details. One can percieve it as either good or bad thing in terms of readability and easiness to understand the tests. I personally find that as soon as you got what `testIt` does, it's easier and faster to read all the tests. On the other hand, it can be just a bit harder to _start_ reading the tests once you opened the file.

If you don't have tests written in such a way and/or don't like it (yes, even subjectively), I'll refactor. But maybe there are people who are ok with that?

Reply:
Any opinions, clarifications, comments or just votes are super-welcome. This helps me a lot with learning the codebase faster.

Reply:
"Newspaper metaphor" with fn declarations?

Reply:
`testIt` helper?